feature/omptarget

jxy opened the pull request:
The OpenMP target backend here is still a work in progress.  We welcome any suggestions.

As of now this port uses a few Intel extensions, contains hacks specifically for Intel architectures, and it only works on Intel GPUs.

For a quick test, try
```
cmake\
        -DCMAKE_BUILD_TYPE=RELEASE\
        -DQUDA_TARGET_TYPE=OMPTARGET\
        -DQUDA_DOWNLOAD_USQCD=on\
        -DQUDA_QMP=on\
        -DQUDA_QIO=on\
        -DQUDA_DIRAC_DEFAULT_OFF=on\
        -DQUDA_DIRAC_STAGGERED=on\
        -DQUDA_PRECISION=8\
        -DQUDA_RECONSTRUCT=4\
        -DQUDA_FAST_COMPILE_REDUCE=on\
        -DQUDA_FAST_COMPILE_DSLASH=on\
        -DQUDA_BUILD_NATIVE_LAPACK=off\
        -DCMAKE_CXX_COMPILER=mpic++\
        -DCMAKE_C_COMPILER=mpicc\
        ../quda
```

jxy committed:
omp: fix reduction by properly initializing omp_priv

jxy committed:
omp: allow program to run when offloading is disabled

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: fix last merge

jxy committed:
omp: ignore *_ctest

jxy committed:
omp: add explicit specialization for zero()

jxy committed:
omp: add more static functions for omp declare reduction [WIP]

jxy committed:
omp: use global variable in omp target for emulating 3D kernels

jxy committed:
omp: use multiple parallel regions in target teams region for block-level reduction

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: add QUDA_RT_CONSTS in coarse_op_kernel for the last merge

jxy committed:
omp: remove unused j in reduction_kernel.h

jxy committed:
omp: fix diagnostic output in reductions

jxy committed:
omp: revert blas_test debug

jxy committed:
omp: add a few commented out printf in reduction

jxy committed:
omp: copy kernel arg to stack to for arg modifying kernels like caxpyXmazMR

jxy committed:
Revert "omp: workaround rng"

This reverts commit 4a31920d148ff54ea22ab5d335456ecf716c7831.

jxy committed:
omp: use rocrand_mrg32k3a for omp target

jxy committed:
omp: better handling of memcpyDefault

jxy committed:
omp: update default device parameters

jxy committed:
omp: try allocator(omp_pteam_mem_alloc) for shared memory [WIP]

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: only warning if no device

jxy committed:
omp: add omp_init/reduce to MomUpdate

jxy committed:
omp: update ompwip functions

jxy committed:
omp: remove cpu side debug print in reduce_helper

jxy committed:
omp: remove debug print in tunable_nd/reduction

jxy committed:
omp: remove debug print in cuda_color_spinor_field

jxy committed:
omp: target/kernel cast pointer to void* before memcpy

jxy committed:
omp: target/math_helper updates and uses generic versions

jxy committed:
omp: target/quda_api.cpp fix memset and add 2D versions

jxy committed:
omp: update malloc debug print

jxy committed:
omp: try turning off shared in dslash_domain_wall_m5

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: fix index_helper

jxy committed:
omp: cleanup malloc and device, add event timing

jxy committed:
omp: remove debug print in kernels

jxy committed:
omp: update math_helper

jxy committed:
omp: remove debug print in tunable_nd

jxy committed:
omp: test kernel launches to make sure the runtime allows the specified parameters

jxy committed:
tunable: correctly respect constraints on launch parameters

- update param.block/grid before calling Tunable::advanceBlockDim(param)
- add test to make sure nthreads <= device::max_threads_per_block()

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: clean up OMP WIP print messages

jxy committed:
omp: reduce constexpr max_block_size to 512

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: incorporate kernel launch changes

jxy committed:
omp: temporary workaround for compiler bugs

jxy committed:
omp: extend omptarget tuning parameters

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
omp: remove unused qudaLaunchKernel

jxy committed:
omp: add declaration

jxy committed:
omp: fix verbose timing

jxy committed:
omp: fix qudaSetupLauchParam's namespace

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
Merge branch 'feature/generic_kernel' into omp

jxy committed:
omp: omptarget conforms to cuda from last merge

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
omp: finish merging target changes

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
Revert "omp: temporary workaround for compiler bugs"

This reverts commit 40fc8a60336a4bbdb2a1ab2f91d51a105c76dc42.

jxy committed:
omp: fix return type of __uint_as_float

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
Revert "omp: try allocator(omp_pteam_mem_alloc) for shared memory [WIP]"

This reverts commit 736600da563cae6c658738853e5e28fdee051cc3.

jxy committed:
Revert "omp: try turning off shared in dslash_domain_wall_m5"

This reverts commit c08b764160c0877fb7cacc17475dd658e96b46c3.

jxy committed:
omp: only define atomics for limited types

jxy committed:
omp: implement shared local memory using stack array inside teams

jxy committed:
omp: implement block/warpreduction pretending to be cub

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
omp: enable cub_helper and block_orthogonalize

jxy committed:
omp: raise warp_size to 16

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
Introduce QUDA_UNROLL for setting #pragma unroll

jxy committed:
omp: use omp atomic capture compare properly

jxy committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into omp

jxy committed:
omp: fix last merge

jxy committed:
omp: replace extra pragma unroll

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: define threadIdx for DW m5 dslash

jxy committed:
merge upstream develop

jxy committed:
omp: fix last merge and clean up target files

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: save unknown pointers in get_pointer_location

jxy committed:
omp: introduce defines for unknown device properties

jxy committed:
omp: back out change to maxGridSize

jxy committed:
replace remaining pragma unroll with QUDA_UNROLL

jxy committed:
omp: prepare for cmake (not cmake ready yet)

jxy committed:
omp: fix quda_arch.h

jxy committed:
omp: remove extra omp.h includes

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: include math_helper in blas_helper

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: update reducer with generic block reduce

jxy committed:
omp: fix get_pointer_location with a hack

jxy committed:
omp: update from cuda (mostly formatting)

jxy committed:
omp: workaround issues with omp barrier

omp barrier requires all threads in the team to participate.
This commit makes sure that happens with
1. reduce() calls BlockReduce in all threads in the team with last block done;
2. qudaLaunchKernel() rejects threads parameters that could result in threads have no work.

jxy committed:
omp: reorder reduce() so all threads call BlockReduce at a single place

jxy committed:
omp: use "omp for" for block reduce

jxy committed:
omp: use explicit memcpy for arg in block reduction

jxy committed:
omp: cmake works

jxy committed:
omp: use our own implementation of MRG32k3a

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: revert generic/fast_intdiv.h to upstream version

mathiaswagner commented:
Jenkins: Can one of the admins verify this patch?

jxy commented:
This is not ready for merge yet.  Just list here for interested people.

maddyscientist commented:
Great to get this up as a draft PR @jxy  😄 

What compilers have you tested this with?

jxy commented:
It currently only works with Intel's.  More information here: https://www.intel.com/content/www/us/en/develop/documentation/oneapi-gpu-optimization-guide/top/openmp-offloading-intro/openmp-compile-and-run.html

maddyscientist commented:
So are you using Intel specific extensions, or is it that other compilers are lacking features?  Curious to know what is missing, for example, with NVIDIA's OMP compiler.

jxy commented:
There are three reasons.

1. QUDA's `mapped_malloc` currently uses `omp_target_alloc_shared`, which is an Intel extension.
2. Different OpenMP implementation may have different interpretation of the specification, and I spent most of my effort on Intel's implementation.  I haven't tried Nvidia's OMP compiler.  Last I tried with llvm (v12 and v13) on Nvidia GPUs (manually copy memories for mapped_alloc), there were issues with atomic, as well as these two bug reports:
   - llvm/llvm-project#51447
   - llvm/llvm-project#51451
3. There are dirty hacks in the code definitely waiting for better solutions:
   a. get pointer location (required for qudaMemcpyDefault) https://github.com/jxy/quda/blob/6329d5735394736dd27289791cbbd5636bd78098/lib/targets/omptarget/malloc.cpp#L652-L660
   b. a single global address for shared memory per team https://github.com/jxy/quda/blob/6329d5735394736dd27289791cbbd5636bd78098/include/targets/omptarget/kernel.h#L173-L174

jxy committed:
omp/mrg32k3a: use numeric_limits::min for TINY

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: update function types in comm_target.cpp

jxy committed:
omp: reworked shared cache system, replaced single address passing

jxy committed:
omp: allocate global memory for shared local memory

jxy committed:
omp: fix min/max for arrays

jxy committed:
omp: cmake allows larger parameters for testing

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: disable simulation of get_num_threads in target

jxy committed:
omp: use firstprivate on teams to get thread private copy (non-stardard behavior?)

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: update reductions to upstream API

Note: some tests fail likely due to unidentified race condition.

jxy committed:
omp: use large shared memory size (in global mem space) by default

jxy committed:
omp: thread/block_idx use uint

jxy committed:
omp: check shared cache allocation

jxy committed:
omp: fix race condition in omp for; replace omp for with manual reductions

jxy committed:
omp: test implementations of block reduce

jxy committed:
omp: using share memory for warp_combine

jxy committed:
omp: change the constraint on qudaLaunchKernel a bit

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: update math_helper

jxy committed:
omp: block_reduce array element iteratively with limited shared mem

jxy committed:
omp: block_reduce split large arrays into smaller ones

jxy committed:
omp: use omp_target_alloc_host for pinned/mapped malloc

jxy committed:
omp: clean up kernel launch

jxy committed:
omp: reduce max_kernel_arg_size to 2048

jxy committed:
omp: use a single buffer for arg and extern in header

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: clean up malloc

jxy committed:
omp: remove device::get_arg, directly passing arg pointer to kernels as an argument

jxy committed:
omp: use constant_kernel_arg for coarse_op

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: fix max_nthr type in warp_collective

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: use MKL cgetrf/i and ?gemm, with the 32 bit integer interface

jxy committed:
omp: reinstate #pragma unroll

jxy committed:
omp: remove some unnecessary QUDA_RT_CONSTS

jxy committed:
omp: add constant_kernel_arg.h for dslash5_domain_wall

jxy committed:
omp: use host pinned memory for global device var

jxy committed:
Revert "omp: use constant_kernel_arg for coarse_op"

This reverts commit 3d434c5f20900521faa66d8687f5f6f848e252a6.

jxy committed:
Revert "omp: add constant_kernel_arg.h for dslash5_domain_wall"

This reverts commit 24af9efa4a8faf9890ab93b4f797a1617af243f6.

jxy committed:
Revert "omp: reduce max_kernel_arg_size to 2048"

This reverts commit f6392616c171f9e5d418156786c9ac2f393188d1.

jxy committed:
blas_core: explicitly static assert kernel arg for caxpyxmazMR_

jxy committed:
kernel_param: force use_kernel_arg when larger than 1

Introduce int use_kernel_arg in BlasFunctor and caxpyxmazMR_,
and force use_kernel_arg with caxpyxmazMR_::use_kernel_arg = 2.

jxy committed:
omp: reduce max_kernel_arg_size down to 64

jxy committed:
omp: correct setup launch param behavior

jxy committed:
cmake: use C++20 by default

jxy committed:
array: remove default for c++20 compatibility

jxy committed:
omp: introducing requires_threads_sync in some kernel args to avoid hanging threads

We can't have threads exits early in openmp and use barrier in the rest of the threads.
OpenMP requires all the threads in a team call participate in a barrier.
In order to help the kernel launcher to filter out potentially hanging kernels,
some kernel args gets a static constexpr ThreadsSync requires_threads_sync,
which tells the kernel launcher if any of the x/y/z thread blocks has to be
divisible to the total threads.

jxy committed:
omp: reserve a fixed amount of global mem to emulate SLM with env QUDA_OMPTARGET_SLM_RESERVE

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: only filter out by arg.threads%tp.block

jxy committed:
omp: fix reserve mem size printf

jxy committed:
omp: tag init_arg and rngArg with ThreadsSyncNo

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'me/use_kernel_arg' into omp

jxy committed:
omp: update reduce_helper with use_kernel_arg

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'me/use_kernel_arg' into omp

jxy committed:
omp: follow use_kernel_arg changes

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp/mkl: use strided API; add zgetrf/i; use 64bit for large batch

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
Merge remote-tracking branch 'origin/develop' into omp

jxy committed:
omp: add pragma once to thread_array.h

jxy committed:
Merge branch 'use_kernel_arg' into omp

jxy committed:
omp: remove our copy of mrg32k3a, use generic

jxy committed:
omp: mark launch_param declare target

jxy committed:
omp/blas_lapack_mkl: use 5.1 standard omp dispatch instead of Intel extension

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: declare target some static constexpr

jxy committed:
color_spinor_pack: tag PackGhostArg with require thread sync X

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: static_assert enough shared cache for shared_atomic VUV

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: follow upstream changes to shared_memory_cache

jxy committed:
omp: set pragma unroll threshold; teach cmake to leave it be

maddyscientist committed:
Fix initialization bug for test_split_grid

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: remove pragma-unroll-threshold from default options

jxy committed:
Merge branch 'hotfix/init_stag_test_link' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: better control of thread limit

jxy committed:
omp: optinally build omp target with jit

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: update backend following upstream

jxy committed:
omp: fix sizeof

A declaration of a parameter as “array of type” shall
be adjusted to “qualified pointer to type”.

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Revert "omp: declare target some static constexpr"

This reverts commit 7a9705d80aefa4405fc8038d013eff3a0dbc0df4.

jxy committed:
omp: make sure extra args/spaces does not break version string

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: update target specific code to conform to upstream

jxy committed:
omp: tests/cmake WA: ignore unresolved symbols from device functions in tune_test

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: update target api following upstream

jxy committed:
omp: use omp groupprivate for slm

- removed global vars in kernel
- use simple array for thread_array

jxy committed:
omp: add option for fixed simd16

jxy committed:
omp: provide variants of kernel launching methods

jxy committed:
omp: update env names and compilation params

jxy committed:
omp: avoid direct char access in memset, use float when possible

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: remove per_kernel compilation and add huge-device-code options

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: update following upstream

jxy committed:
omp: provide QUDA_OMPTARGET_THREAD_ARRAY_SLM to move thread_array location

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: include <utility>

jxy committed:
omp: QUDA_OMPTARGET_DEBUG only CPU side for now

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge branch 'unpack_fix' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: fix omptarget after merge

jxy committed:
Merge remote-tracking branch 'upstream/hotfix/complex_template' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: work around a compiler bug

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: add empty device::get_state

jxy committed:
BlockKernel2D_host: create block inside omp parallel

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: use memcpy for vector_load/store

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omptarget/reduce_helper.h: atomic read the partial results

omp flush doesn't seem to do what sycl::atomic_fence does.
the atomic read seems to enforce the sequential consistency for the partial array
we will revisit this once we have got better omp support from vendors

jxy committed:
Merge remote-tracking branch 'upstream/develop' into omp

jxy committed:
omp: omptarget/atomic_helper: add atomic_read for complex<T>
